File: frontend/lib/api/jobs.ts
export type JobStatus =
  | "queued"
  | "downloading"
  | "downloaded"
  | "separating_stems"
  | "stems_separated"
  | "transcribing"
  | "transcribed"
  | "mapping_tempo"
  | "tempo_mapped"
  | "failed";

export interface Job {
  id: string;
  url: string;
  status: JobStatus;
  audio_path?: string | null;
  drums_path?: string | null;
  accompaniment_path?: string | null;
  event_count?: number | null;
  tempo_bpm?: number | null;
  error?: string | null;
}

export type DrumInstrument =
  | "kick"
  | "snare"
  | "hihat_closed"
  | "hihat_open"
  | "crash"
  | "ride"
  | "tom_low"
  | "tom_mid"
  | "tom_high";

export interface AnalysisEvent {
  id: string;
  time: number;
  instrument: DrumInstrument;
  confidence: number | null;
  provenance: string | null;
  measure: number | null;
  beat: number | null;
  subdivision: number | null;
}

export interface Beat {
  source_time: number;
  measure: number;
  beat: number;
  is_downbeat: boolean;
  confidence: number | null;
}

export interface Analysis {
  tempo_bpm: number;
  events: AnalysisEvent[];
  beats: Beat[];
}

async function parseResponse<T>(response: Response): Promise<T> {
  const body = await response.json();

  if (!response.ok) {
    throw new Error(body.detail ?? "Request failed");
  }

  return body;
}

export async function createJob(baseUrl: string, url: string): Promise<Job> {
  const response = await fetch(`${baseUrl}/api/jobs`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ url }),
  });

  return parseResponse<Job>(response);
}

export async function getJob(baseUrl: string, id: string): Promise<Job> {
  const response = await fetch(`${baseUrl}/api/jobs/${id}`);

  return parseResponse<Job>(response);
}

export async function getAnalysis(baseUrl: string, id: string): Promise<Analysis> {
  const response = await fetch(`${baseUrl}/api/jobs/${id}/analysis`);

  return parseResponse<Analysis>(response);
}

File: backend/app/jobs.py
import dataclasses
import uuid
from dataclasses import dataclass
from datetime import UTC, datetime
from enum import Enum

from app.transcription import DrumEvent


class JobStatus(str, Enum):
    QUEUED = "queued"
    DOWNLOADING = "downloading"
    DOWNLOADED = "downloaded"
    SEPARATING_STEMS = "separating_stems"
    STEMS_SEPARATED = "stems_separated"
    TRANSCRIBING = "transcribing"
    TRANSCRIBED = "transcribed"
    FAILED = "failed"


@dataclass(frozen=True)
class Job:
    id: str
    url: str
    status: JobStatus
    created_at: datetime
    audio_path: str | None = None
    drums_path: str | None = None
    accompaniment_path: str | None = None
    events: list[DrumEvent] | None = None
    error: str | None = None


class JobStore:
    def __init__(self) -> None:
        self._jobs: dict[str, Job] = {}

    def create(self, url: str) -> Job:
        job = Job(
            id=str(uuid.uuid4()),
            url=url,
            status=JobStatus.QUEUED,
            created_at=datetime.now(UTC),
        )
        self._jobs[job.id] = job
        return job

    def get(self, job_id: str) -> Job | None:
        return self._jobs.get(job_id)

    def update(self, job_id: str, **changes: object) -> Job:
        current = self._jobs[job_id]
        updated = dataclasses.replace(current, **changes)
        self._jobs[job_id] = updated
        return updated
